src/spatialdata_plot/pl/utils.py

```diff
@@ -1031,3 +1051,186 @@ def _mpl_ax_contains_elements(ax: Axes) -> bool:
+
+    Valid cs are cs that contain elements to be rendered:
+    1. In case the user specified elements:
+        all cs that contain at least one of those elements
+    2. Else:
+        all cs that contain at least one element that should
+        be rendered (depending on whether images/points/labels/...
+        should be rendered)
+    """
+    cs_mapping = _get_coordinate_system_mapping(sdata)
+    valid_cs = []
+    for cs in coordinate_systems:
+        if (len(elements) > 0 and any(e in elements for e in cs_mapping[cs])) or (
+            len(elements) == 0
+            and (
+                (len(sdata.images.keys()) > 0 and render_images)
+                or (len(sdata.labels.keys()) > 0 and render_labels)
+                or (len(sdata.points.keys()) > 0 and render_points)
+                or (len(sdata.shapes.keys()) > 0 and render_shapes)
+            )
+        ):  # not nice, but ruff wants it (SIM114)
+            valid_cs.append(cs)
+        else:
+            logging.info(f"Dropping coordinate system '{cs}' since it doesn't have relevant elements.")
+    return valid_cs
+
+
+def _rasterize_if_necessary(
+    image: SpatialImage,
+    dpi: float,
+    width: float,
+    height: float,
+    coordinate_system: str,
+    extent: dict[str, tuple[float, float]],
+) -> SpatialImage:
+    """Ensure fast rendering by adapting the resolution if necessary.
+
+    A SpatialImage is prepared for plotting. To improve performance, large images are rasterized.
+
+    Parameters
+    ----------
+    image
+        Input spatial image that should be rendered
+    dpi
+        Resolution of the figure
+    width
+        Width (in inches) of the figure
+    height
+        Height (in inches) of the figure
+    coordinate_system
+        name of the coordinate system the image belongs to
+    extent
+        extent of the (full size) image. Must be a dict containing a tuple with min and
+        max extent for the keys "x" and "y".
+
+    Returns
+    -------
+    SpatialImage
+        Spatial image ready for rendering
+    """
+    has_c_dim = len(image.shape) == 3
+    if has_c_dim:
+        y_dims = image.shape[1]
+        x_dims = image.shape[2]
+    else:
+        y_dims = image.shape[0]
+        x_dims = image.shape[1]
+
+    target_y_dims = dpi * height
+    target_x_dims = dpi * width
+
+    # TODO: when exactly do we want to rasterize?
+    do_rasterization = y_dims > target_y_dims + 100 or x_dims > target_x_dims + 100
+    if x_dims < 2000 and y_dims < 2000:
+        do_rasterization = False
+
+    if do_rasterization:
+        # TODO: do we want min here?
+        target_unit_to_pixels = min(target_y_dims / y_dims, target_x_dims / x_dims)
+        image = rasterize(
+            image,
+            ("y", "x"),
+            [extent["y"][0], extent["x"][0]],
+            [extent["y"][1], extent["x"][1]],
+            coordinate_system,
+            target_unit_to_pixels=target_unit_to_pixels,
+        )
+        logging.info(
```

Comment:
I would remove this, I think it's enough for the user to experience fast performance and then highlight the performance improvement in a notebook.